websocket.go

```diff
@@ -21,6 +22,72 @@ import (
 	graceful "gopkg.in/tylerb/graceful.v1"
 )
 
+const certificateReloaderLeeway = 1 * time.Hour
+
+// CertificateReloader takes care of reloading a TLS certificate when
+// requested.
+type CertificateReloader struct {
+	sync.RWMutex
+	cert     *tls.Certificate
+	certPath string
+	keyPath  string
+}
+
+// NewCertificateReloader takes two file paths as parameter that contain
+// the certificate and the key data to create an automatic reloader. It will
+// try to read again the files when the certificate is almost expired.
+func NewCertificateReloader(certPath, keyPath string) (*CertificateReloader, error) {
+	loader := &CertificateReloader{
+		certPath: certPath,
+		keyPath:  keyPath,
+	}
+
+	return loader, loader.reload()
+}
+
+func (cr *CertificateReloader) reload() error {
+	newCert, err := tls.LoadX509KeyPair(cr.certPath, cr.keyPath)
+	if err != nil {
+		return err
+	}
+
+	cr.Lock()
```

Comment:
What if the pre-condition to calling reload() is that you hold the read lock, and then reload() takes responsibility for releasing the read lock and getting the write lock if it needs to? That means one more lock/unlock in NewCertificateReloader but then simplified code in GetCertificateFunc().

The certificate function is called 1 time for every incoming TLS connection, so it really needs to be as lightweight as possible.